controllers/fencer/reconciler.go

```diff
@@ -0,0 +1,206 @@
+
+	// Don't allow an overly-agressive interval to break the backend.
+	if r.pollInterval < minPollInterval {
+		r.log.Info("resetting poll interval to minimum", "interval", minPollInterval)
+	}
+
+	// Periodically populate the cache by sending StorageOS API nodes to the
+	// event source.
+	go r.pollNodes(ctx, src)
+
+	// Create an index on the Pod's node name.
+	if err := mgr.GetFieldIndexer().IndexField(ctx, &corev1.Pod{}, "spec.nodeName", func(rawObj client.Object) []string {
+		pod := rawObj.(*corev1.Pod)
+		return []string{pod.Spec.NodeName}
+	}); err != nil {
+		return err
+	}
+
+	// Create an index on the VolumeAttachment's node name.
+	if err := mgr.GetFieldIndexer().IndexField(ctx, &storagev1.VolumeAttachment{}, "spec.nodeName", func(rawObj client.Object) []string {
+		va := rawObj.(*storagev1.VolumeAttachment)
+		return []string{va.Spec.NodeName}
+	}); err != nil {
+		return err
+	}
+
+	return builder.ControllerManagedBy(mgr).
+		Named("fencer").
+		WithOptions(controller.Options{MaxConcurrentReconciles: workers}).
+		WithSource(src).
+		WithEventHandler(eventHandler).
+		Complete(r)
+}
+
+// pollNodes periodically retrieves the StorageOS nodes and sends them over the
+// source event channel.  It is blocking and is intended to run in a goroutine,
+// stopping when the context is cancelled.
+func (r *Reconciler) pollNodes(ctx context.Context, src chan event.GenericEvent) {
+	ticker := time.NewTicker(r.pollInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			tr := otel.Tracer("fencer")
+			ctx, span := tr.Start(ctx, "fencing poller")
+			r.log.V(5).Info("polling for external node health")
+
+			nodes, err := r.api.ListNodes(ctx)
+			if err != nil {
+				span.RecordError(errors.Wrap(err, "failed to list nodes"))
+				r.log.Error(err, "failed to list nodes")
+
+				// We should always be able to list nodes.  Reset the api client
+				// if we got an error.
+				r.apiReset <- struct{}{}
+
+				span.End()
+				continue
+			}
+			span.SetAttributes(label.Int("nodes", len(nodes)))
+			for _, node := range nodes {
+				src <- event.GenericEvent{
+					Object: node,
+				}
+			}
+			span.SetStatus(codes.Ok, "refreshed node cache")
+			span.End()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// Reconcile is part of the main node health reconcile loop which aims to move
+// the current state of the cluster closer to the desired state.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	tr := otel.Tracer("fencer")
+	ctx, span := tr.Start(ctx, "node fencer reconcile")
+	span.SetAttributes(label.String("name", req.NamespacedName.Name))
+	defer span.End()
+	log := r.log.WithValues("name", req.NamespacedName)
+
+	// Query the object details from the cache and operate on it.
+	val, found := r.cache.Get(req.NamespacedName.String())
+	if !found {
+		log.Error(ErrNodeNotCached, "skipping node, will retry")
+		span.RecordError(ErrNodeNotCached)
+		return ctrl.Result{}, nil
```

Comment:
The log says that it'll retry, but the result has no requeue set. I think that's fine, the log should not say it'll retry because usually we don't know what kind of event we received, in case of delete event, Get fails and we don't do anything. But in case of this specific controller, it only receives generic events, not create, update or delete events. So an event means something changed and we may not ignore that and actually retry.
This makes me think about the consequences of 1 minute cache expiry time. If due to some reason, we retry here or also below where `fenceNode()` is called, and due to some reason we fail to reconcile within a minute, the cache will be empty and we'll not be able to do anything because we'll not have the event data. Having a longer cache expiry and cleanup time doesn't hurt because cache gets updated for any observed change in objects.

Reply:
I'm now returning an error here, and I've increased the cache expiry to 10 minutes.  An hour might be better?